lib/mocha/parameter_matchers/positional_or_keyword_hash.rb

```diff
@@ -9,14 +10,24 @@ def initialize(value)
         @value = value
       end
 
-      def matches?(available_parameters)
+      def matches?(available_parameters) # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
         parameter, is_last_parameter = extract_parameter(available_parameters)
         return false unless parameter.is_a?(Hash)
+        return false unless parameter == @value
 
-        if is_last_parameter && Mocha.configuration.strict_keyword_argument_matching?
-          return false unless ::Hash.ruby2_keywords_hash?(parameter) == ::Hash.ruby2_keywords_hash?(@value)
+        if is_last_parameter && (::Hash.ruby2_keywords_hash?(parameter) != ::Hash.ruby2_keywords_hash?(@value))
+          return false if Mocha.configuration.strict_keyword_argument_matching?
+
+          if Mocha::RUBY_V27_PLUS
+            if ::Hash.ruby2_keywords_hash?(@value)
+              deprecation_warning("Expected keyword arguments (#{mocha_inspect}), but received positional hash (#{parameter.mocha_inspect}).")
+            else
+              deprecation_warning("Expected positional hash (#{mocha_inspect}), but received keyword arguments (#{parameter.mocha_inspect}).")
+            end
```

Comment:
In the meantime, maybe this could be moved into the deprecation_warning method itself? Removes a bit of nesting 😄

Reply:
> In the meantime, maybe this could be moved into the deprecation_warning method itself? Removes a bit of nesting 😄

Addressed in https://github.com/freerange/mocha/pull/563/commits/5c33923fe1eb3bbeb39ab72200de6bc5a6a835fa.